packages/server/lib/cloud/protocol.ts

```diff
@@ -0,0 +1,77 @@
+import fs from 'fs-extra'
+import { NodeVM } from 'vm2'
+import Debug from 'debug'
+import CDP from 'chrome-remote-interface'
+import type { ProtocolManager } from '@packages/types'
+
+interface AppCaptureProtocolInterface {
+  setupProtocol (url?: string): Promise<void>
+  connectToBrowser (options: any): void
+  beforeSpec (spec: any): void
+  afterSpec (): void
+  beforeTest (test: any): void
+}
+
+const debug = Debug('cypress:server:protocol')
+
+const setupProtocol = async (url?: string): Promise<AppCaptureProtocolInterface | undefined> => {
+  let script: string | undefined
+
+  // TODO: We will need to remove this option in production
+  if (process.env.CYPRESS_LOCAL_PROTOCOL_PATH) {
+    script = await fs.readFile(process.env.CYPRESS_LOCAL_PROTOCOL_PATH, 'utf8')
+  } else if (url) {
+    // TODO: Download the protocol script from the cloud
+  }
+
+  if (script) {
+    const vm = new NodeVM({
+      console: 'inherit',
+      sandbox: { Debug, CDP },
+    })
+
+    const { AppCaptureProtocol } = vm.run(script)
+
+    return new AppCaptureProtocol()
+  }
+
+  return
+}
+
+class ProtocolManagerImpl implements ProtocolManager {
+  private protocol: AppCaptureProtocolInterface | undefined
+
+  protocolEnabled (): boolean {
+    return !!this.protocol
+  }
+
+  async setupProtocol (url?: string) {
+    debug('setting up protocol')
```

Comment:
Would it be helpful to include the `url` in the debug log?

Reply:
[`fcb6c6f` (#26080)](https://github.com/cypress-io/cypress/pull/26080/commits/fcb6c6f9833d27a6958eb0a8dbde8449fabc6447#diff-fae7886d3dd070e07a8b39ac5718534511764b7f03bc24c9726f82977b53a225)